doc/sphinx/proofs/writing-proofs/equality.rst

```diff
@@ -530,10 +530,8 @@ which reduction engine to use.  See :ref:`type-cast`.)  For example:
        unfolding is applied to all constants that are not listed.
        Notice that the ``delta`` doesn't apply to variables bound by a let-in
        construction inside the term itself (use ``zeta`` to inline these).
-       Opaque constants are never unfolded except by :tacn:`vm_compute` and
-       :tacn:`native_compute`
-       (see `#4476 <https://github.com/coq/coq/issues/4476>`_ and
-       :ref:`controlling-the-reduction-strategies`).
+       Opaque constants are not unfolded by most tactics
+       (see :ref:`controlling-the-reduction-strategies`).
```

Comment:
```suggestion
       Constants marked opaque with :cmd:`Opaque` are not unfolded by most reduction tactics
       (see :ref:`controlling-the-reduction-strategies`).
```

Just making sure to distinguish constants made `Opaque` from truly opaque constants (defined with `Qed` or sealed using modules). And being more explicit about which tactics are affected. (Many tactics use unification which will happily ignore `Opaque`.)

Reply:
The Opaque command should mention this distinction.  Unification only pays attention to Qed opaqueness?  How can I create an example showing that?

How many tactics use unification?  Elsewhere I mention apply and reflexivity ignore Opaque.

How do you seal a constant with a module?  This is not sufficient:

![image](https://github.com/user-attachments/assets/a6ef5c0f-efac-4910-ae8d-b5fcf8a10e2c)

nor this:

![image](https://github.com/user-attachments/assets/c9ab0609-a68d-4190-a846-c46179da9525)

Reply:
> The `Opaque` command should mention this distinction. Unification only pays attention to Qed opaqueness? How can I create an example showing that?

```coq
Definition one : nat := 1.
Opaque one.
Theorem thm : 1 = one.
Proof.
  Succeed (lazymatch goal with |- ?l = ?r => unify l r end). (* tactic unification (most standard Coq tactics) *)
  Succeed refine (eq_refl).                                  (* evarconv unification (pretyper, ssreflect tactics) *)
  reflexivity.
Qed.
```

> How many tactics use unification? Elsewhere I mention apply and reflexivity ignore Opaque.

Almost everything uses unification in one way or another. My mental model is that `Opaque` only affects reduction tactics and anything else that somebody thought should honor `Opaque`. No example comes to mind, though.

> How do you seal a constant with a module? This is not sufficient:

You need module types for module sealing. I should have been more explicit. Sorry about that.

```coq
Module Type MT.
  Axiom one : nat.
End MT.
Module M : MT.
  Definition one : nat := 1.
End M.
Theorem thm : 1 = M.one. Fail reflexivity. Abort.
```

Reply:
Thanks.

Interestingly, dropping in `Transparent M.one` before the `Theorem` is executed but has no effect.  Also the `About` command doesn't report anything about the opacity status.

Reply:
> > Elsewhere I mention apply and reflexivity ignore Opaque.
> 
> Almost everything uses unification in one way or another. My mental model is that `Opaque` only affects reduction tactics and anything else that somebody thought should honor `Opaque`.

`Opaque` affects all conversion, including `reflexivity`, `apply`, `Qed`.  But it's treated differently (more strictly) by reduction tactics.  `Opaque` is roughly the same as `Strategy` infinity.  You can see this if you try to prove `fact 10 = 10 * fact 9` (or the symmetric version).  One of these will go instantaneously by `reflexivity`, unless you say `Opaque fact`, in which case it'll take a very long time.

Reply:
The `fact` expression you give takes .448 seconds vs 4.491 seconds for the symmetric expression, the latter much quicker than I expected.  Two thoughts:
- Should the section on conversion give this example and recommend setting computation-intensive items to `Opaque`?  Is it a simple thing everyone should do routinely?  Or can it be tricky to work with.
- Would there be high value in making `reflexivity` do something smarter?  I suspect that it maynot be simple even if it is conceptually possible.

> `Opaque` affects all conversion, including `reflexivity`, `apply`, `Qed`.

You may be right, but I wasn't to convince myself that it's true for `apply`.  Would be nice to elaborate/document this fully at some point.  Let's not do that now.